### PR TITLE
fix: Key the address book by chain ID instead of network ID

### PR DESCRIPTION
**Pull request by Gudahtt**
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/coding_guidelines/CODING_GUIDELINES.md)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add the appropiate QA label when dev review is completed
    - `needs-qa`: PR requires manual QA.
    - `No QA/E2E only`: PR does not require any manual QA effort. Prior to merging, ensure that you have successful end-to-end test runs in Bitrise. 
    - `Spot check on release build`: PR does not require feature QA but needs non-automated verification. In the description section, provide test scenarios. Add screenshots, and or recordings of what was tested.
5. Add `QA Passed` label when QA has signed off (Only required if the PR was labeled with `needs-qa`)
6. Add your team's label, i.e. label starting with `team-` (or `external-contributor` label if your not a MetaMask employee)

**Description**

The `AddressBookController` stores addresses by chain ID, but we had been storing and retrieving them using the network ID instead in many places. It has been updated to consistently use the chain ID for all address book access.

Existing address book entries have been migrated to be grouped by chain ID. For all known networks, we attempt to map the network ID to a chain ID that exists as a locally configured or built-in network. In cases where multiple matches are found, the entries are duplicated on each matching chain. Address book entries that don't correspond with any local networks are discarded.

**Screenshots / Recordings**

### Before

In this recording, I create a contact on ETC and show that it's also suggested on Ethereum Mainnet

https://recordit.co/KrZMjEYKb4

### After

In this recording, I am running a test build on this branch with the same wallet state as for the "Before" recording, so the test ETC contact entry is still there. I show that it was duplicated onto both Ethereum Mainnet and ETC. Then I create a new contact entry and show that it's correctly saved just for the current chain.

https://recordit.co/AIzatKvKXr

**Issue**

This relates to https://github.com/MetaMask/mobile-planning/issues/1226

**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

### CODE REVIEW

File: app/components/UI/ApproveTransactionReview/AddNickname/index.tsx [hunk 1]
```diff
@@ -75,17 +73,16 @@ const AddNickname = (props: AddNicknameProps) => {
   const validateAddressOrENSFromInput = useCallback(async () => {
     const { addressError, errorContinue } = await validateAddressOrENS({
       toAccount: address,
-      // TODO: This parameters is effectively ignored, it should be named `networkId`
-      providerNetwork,
       addressBook,
       identities,
+      // TODO: This parameters is effectively ignored, it should be named `chainId`
```
Comment (Gudahtt) [inline app/components/UI/ApproveTransactionReview/AddNickname/index.tsx:78]: I had missed this when I left the other comment. The same bug affects the chainId parameter as well.

File: app/store/migration-data/amibiguous-networks.json
Comment (Gudahtt) [inline app/store/migration-data/amibiguous-networks.json:1]: This was generated using this script: https://gist.github.com/Gudahtt/7a8a9e452bd2efdc5ceecd93610a25d3

File: app/store/migrations.js [hunk 2]
```diff
@@ -493,6 +499,171 @@ export const migrations = {
...
+   * @param {any} state - Redux state.
+   * @returns Migrated Redux state.
+   */
+  22: (state) => {
+    const networkControllerState =
+      state.engine.backgroundState.NetworkController;
+    const addressBookControllerState =
+      state.engine.backgroundState.AddressBookController;
+
+    if (!isObject(networkControllerState)) {
```
Comment (Gudahtt) [inline app/store/migrations.js:528]: This follows a pattern that we've recently started using for extension migrations, where we exhaustively check for invalid state (for the parts the migration operates on) before changing it. Then any invalid state is registered to Sentry as an error, but doesn't _throw_ an error and doesn't block future migrations. This alerts us to unexpected state without breaking the experience for the user, letting us fix the invalid state issues in a follow-up release

**Comment by Gudahtt** [comment]
@SocketSecurity ignore-all

False positive, the new dependencies detected are not new

**Comment by Gudahtt** [comment]
I have updated the migration to store a record of all ambiguous addresses that have been duplicated across chains. I am awaiting further clarification on the next requirements.

**Comment by Gudahtt** [comment]
>2. Add a warning icon to addresses that had chain conflicts
>3. Add tooltip/new design pattern to verify the chain (I don't know if we have actions on addresses today, e.g. remove, but we need something like this)

These steps are being discussed, so I have marked this as `DO-NOT-MERGE` for now. But aside from that, the PR is ready for review

File: app/store/migrations.js [hunk 3]
```diff
@@ -493,6 +499,188 @@ export const migrations = {
...
+          (chainId) => localChainIds.has(chainId),
+        );
+
+        if (recognizedChainIdCandidates.length > 1) {
+          for (const chainId of recognizedChainIdCandidates) {
+            ambiguousAddressEntries[chainId] = Object.keys(addressEntries);
+          }
+        }
+
+        for (const chainId of recognizedChainIdCandidates) {
```
Comment (Cal-L) [inline app/store/migrations.js:664]: Should be able to consolidate this logic in the same loop as line 659.

Comment (Gudahtt) [reply app/store/migrations.js]: Done in https://github.com/MetaMask/metamask-mobile/pull/7197/commits/64450f70cc78f4ceea89870f5f9be139e00491cb

**Review by Cal-L — DISMISSED** [review]
LGTM. LMK if you need another review once there's more movement on UI dependencies.

**Comment by Gudahtt** [comment]
This has been rebased to resolve conflicts (the migration was moved from 22 to 23 because a migration 22 was added on main)

**Comment by Gudahtt** [comment]
The PR description has been updated with recordings

**Review by Cal-L — APPROVED** [review]
LGTM. Migration works when upgrading from old version to new. Existing contacts show up the same as they did before. Ambiguous addresses are populated in redux. Adding contacts post fix does not cross contaminate across chains. Nice.

**Comment by Cal-L** [comment]
Removing do not merge label for now since we will handle the UI improvement in a separate task. Follow thread here - https://consensys.slack.com/archives/CBW7S9FSN/p1694536315593899

**Review by cortisiko — APPROVED** [review]
I did a manual regression on this PR with an emphasis on the testing contract nicknames functionality: contract nicknames should appear in the address book but not in the send flow. 

✅

**Pull request merged**